{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4876597a",
   "metadata": {},
   "source": [
    "# Mitochondrial Gene Identification\n",
    "\n",
    "This notebook integrates steps to identify mitochondrial genes in a new species using a reference species and analyze expression correlation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e95e7655",
   "metadata": {},
   "source": [
    "## 1. Preprocess GTF to keep longest transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "356f21ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import gffutils\n",
    "import pandas as pd\n",
    "\n",
    "def keep_longest_transcripts(gtf_path, out_gtf, db_path=None):\n",
    "    \"\"\"Keep only the longest transcript for each gene in a GTF file.\"\"\"\n",
    "    db_path = db_path or gtf_path + '.db'\n",
    "    db = gffutils.create_db(gtf_path, dbfn=db_path, force=True,\n",
    "                            keep_order=True, disable_infer_genes=True,\n",
    "                            disable_infer_transcripts=True)\n",
    "    tx_lengths = []\n",
    "    for tx in db.features_of_type('transcript', order_by='start'):\n",
    "        exons = list(db.children(tx, featuretype='exon', order_by='start'))\n",
    "        length = sum(e.end - e.start + 1 for e in exons)\n",
    "        gene_id = tx.attributes['gene_id'][0]\n",
    "        tx_lengths.append((tx.id, gene_id, length))\n",
    "    df = pd.DataFrame(tx_lengths, columns=['transcript_id','gene_id','length'])\n",
    "    longest_tx = df.sort_values('length').drop_duplicates('gene_id', keep='last')\n",
    "    with open(out_gtf, 'w') as out:\n",
    "        for tx_id in longest_tx.transcript_id:\n",
    "            for feature in db.children(tx_id, featuretype='exon', order_by='start', level=1):\n",
    "                out.write(str(feature) + '\n",
    "')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8e5596e",
   "metadata": {},
   "source": [
    "Use the function for mouse and species X:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "818bdc1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# keep_longest_transcripts('mouse.gtf', 'mouse_longest.gtf')\n",
    "# keep_longest_transcripts('speciesX.gtf', 'speciesX_longest.gtf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f5bfbe4",
   "metadata": {},
   "source": [
    "## 2. Find reciprocal best BLAST hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad35e52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%%bash\n",
    "# Build BLAST databases and run reciprocal BLAST searches\n",
    "# Adjust file paths accordingly\n",
    "\n",
    "makeblastdb -in speciesX_longest.fa -dbtype nucl\n",
    "blastn -query mouse_longest.fa -db speciesX_longest.fa        -out mouse_vs_X.tsv -outfmt 6 -evalue 1e-5 -max_target_seqs 1\n",
    "\n",
    "makeblastdb -in mouse_longest.fa -dbtype nucl\n",
    "blastn -query speciesX_longest.fa -db mouse_longest.fa        -out X_vs_mouse.tsv -outfmt 6 -evalue 1e-5 -max_target_seqs 1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e52a8d4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "\n",
    "m_to_x = pd.read_csv('mouse_vs_X.tsv', sep='\t', header=None,\n",
    "                     usecols=[0,1], names=['mouse','X'])\n",
    "x_to_m = pd.read_csv('X_vs_mouse.tsv', sep='\t', header=None,\n",
    "                     usecols=[0,1], names=['X','mouse'])\n",
    "rbh = m_to_x.merge(x_to_m, on=['mouse','X'])\n",
    "mouse_mt = pd.read_csv('mouse_mt_genes.txt', header=None, names=['mouse'])\n",
    "rbh_mt = rbh.merge(mouse_mt, on='mouse')\n",
    "rbh_mt.to_csv('X_mito_candidate.tsv', sep='\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b616873",
   "metadata": {},
   "source": [
    "## 3. Map candidate genes to contigs or chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94113add",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "\n",
    "gtf_cols = ['seqname','source','feature','start','end','score','strand','frame','attribute']\n",
    "gtf_df = pd.read_csv('speciesX_longest.gtf', sep='\t', comment='#', names=gtf_cols)\n",
    "gtf_df['gene_id'] = gtf_df.attribute.str.extract('gene_id \"([^\"]+)\"')\n",
    "\n",
    "cand = pd.read_csv('X_mito_candidate.tsv', sep='\t')\n",
    "result = gtf_df[gtf_df.gene_id.isin(cand['X'])]\n",
    "result[['seqname','gene_id']].drop_duplicates().groupby('seqname')['gene_id']       .apply(list).to_csv('contig_gene_mapping.txt')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c3de597",
   "metadata": {},
   "source": [
    "## 4. Expression correlation in scRNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "225368de",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "adata = sc.read_h5ad('speciesX_scRNA.h5ad')\n",
    "cand = pd.read_csv('X_mito_candidate.tsv', sep='\t')\n",
    "genes = cand['X'].tolist()\n",
    "adata_mt = adata[:, [g for g in genes if g in adata.var_names]]\n",
    "\n",
    "expr = adata_mt.X.toarray() if hasattr(adata_mt.X, 'toarray') else adata_mt.X\n",
    "corr = np.corrcoef(expr.T)\n",
    "\n",
    "sns.clustermap(corr, cmap='vlag', center=0, xticklabels=genes, yticklabels=genes)\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
